File: src/main/kotlin/com/advent2022/puzzle4/Puzzle4.kt
package com.advent2022.puzzle4

import com.advent2021.base.Base

typealias Data = ArrayList<Pair<IntRange, IntRange>>
typealias Solution = Int
typealias Solution2 = Solution

// https://adventofcode.com/2022/day/4
fun main() {
    try {
        val puz = Puzzle4()
        val solution1 = puz.solvePuzzle("inputs.txt", Data())
        println("Solution1: $solution1")

        val solution2 = puz.solvePuzzle2("inputs.txt", Data())
        println("Solution2: $solution2")
    } catch (t: Throwable) {
        t.printStackTrace()
    }
}

fun String.toRange(): IntRange {
    val parts = this.split("-")
    return IntRange(parts[0].toInt(), parts[1].toInt())
}

class Puzzle4 : Base<Data, Solution?, Solution2?>() {
    override fun parseLine(line: String, data: Data) {
        val parts = line.split(",")
        val rangePair = Pair(parts[0].toRange(), parts[1].toRange())
        data.add(rangePair)
    }

    override fun computeSolution(data: Data): Solution {
        return data.sumOf { computeOne(it) }
    }

    private fun computeOne(pair: Pair<IntRange, IntRange>): Int {
        return if (overlapCompletely(pair)) {
            1
        } else {
            0
        }
    }

    private fun overlapCompletely(pair: Pair<IntRange, IntRange>): Boolean {
        val firstStart = pair.first.first
        val firstEnd = pair.first.last
        val secondStart = pair.second.first
        val secondEnd = pair.second.last
        return ((firstStart <= secondStart && firstEnd >= secondEnd) ||
            secondStart <= firstStart && secondEnd >= firstEnd)
    }

    override fun computeSolution2(data: Data): Solution2 {
        return data.sumOf { computeSecond(it) }
    }

    private fun computeSecond(pair: Pair<IntRange, IntRange>): Int {
        return if (overlapCompletely(pair) || overlapPartial(pair)) {
            1
        } else {
            0
        }
    }

    private fun overlapPartial(pair: Pair<IntRange, IntRange>): Boolean {
        val firstRange = pair.first
        val secondRange = pair.second
        return firstRange.contains(secondRange.first) || firstRange.contains(secondRange.last)
                || secondRange.contains(firstRange.first) || secondRange.contains(firstRange.last)
    }
}

File: src/main/ruby/advent2023/Puzzle4.rb
load 'base.rb'
require 'Set'

module Puzzle4
  class Puzzle
    include Base

    def main
      solution1 = self.solve_puzzle("puzzle4/inputs.txt", Array.new) do |line, data|
        data << ScratchCard.new(line)
      end
      puts "Solution1: #{solution1}"

      solution2 = self.solve_puzzle2("puzzle4/inputs.txt", Array.new) do |line, data|
        data << ScratchCard.new(line)
      end
      puts "Solution2: #{solution2}"
    end

    def compute_solution(data)
      data.sum do |card|
        matching = card.matching_numbers.size
        (matching == 0 ? 0 : 2.pow(matching - 1))
      end
    end

    def compute_solution2(data)
      map = Array.new(data.size, 1)
      data.each do |card|
        num = card.card_num
        matching = card.matching_numbers.size
        (num .. num + matching - 1).each do |n|
          map[n] = map[n] + map[num - 1]
        end
      end
      map.sum
    end
  end

  class ScratchCard
    attr_reader :card_num

    @winning_numbers = []
    @picked_numbers = []
    @card_num

    def initialize(str)
      str = str["Card ".length..]
      @card_num = str.to_i
      str = str[str.index(':')+1..]
      parts = str.split("|")
      @winning_numbers = parts[0].split(" ").map { |s| s.to_i }
      @picked_numbers = parts[1].split(" ").map { |s| s.to_i }
    end

    def matching_numbers
      winning = Set.new @winning_numbers
      @picked_numbers.filter {|n| winning.member? n}
    end
  end
end

p = Puzzle4::Puzzle.new
p.main